packages/app/src/background/services/credentials/index.ts

```diff
@@ -0,0 +1,176 @@
+    this.lockService = LockerService.getInstance();
+    this.historyService = HistoryService.getInstance();
+    this.notificationService = NotificationService.getInstance();
+  }
+
+  static getInstance(): VerifiableCredentialsService {
+    if (!VerifiableCredentialsService.INSTANCE) {
+      VerifiableCredentialsService.INSTANCE = new VerifiableCredentialsService();
+    }
+
+    return VerifiableCredentialsService.INSTANCE;
+  }
+
+  addVerifiableCredential = async (verifiableCredentialJson: string): Promise<boolean> => {
+    if (!verifiableCredentialJson || typeof verifiableCredentialJson !== "string") {
+      return false;
+    }
+
+    const verifiableCredential = parseVerifiableCredentialFromJson(verifiableCredentialJson);
+    if (verifiableCredential === null) {
+      return false;
+    }
+    return this.insertVerifiableCredentialIntoStore(verifiableCredential);
+  };
+
+  getAllVerifiableCredentials = async (): Promise<VerifiableCredential[]> =>
+    this.getVerifiableCredentialsFromStore()
+      .then((credentials) => Array.from(credentials.values()))
+      .then((credentialsArray) =>
+        credentialsArray.map((credential) => parseVerifiableCredentialFromJson(credential) as VerifiableCredential),
+      );
+
+  deleteVerifiableCredential = async (credentialId: string): Promise<boolean> => {
+    if (!credentialId || typeof credentialId !== "string") {
+      return false;
+    }
+
+    const credentials = await this.getVerifiableCredentialsFromStore();
+    if (!credentials.has(credentialId)) {
+      return false;
+    }
+
+    credentials.delete(credentialId);
+    await this.writeVerifiableCredentials(credentials);
+    await this.historyService.trackOperation(OperationType.DELETE_VERIFIABLE_CREDENTIAL, {});
+    await this.notificationService.create({
+      options: {
+        title: "Verifiable Credential deleted.",
+        message: `Credential Id: ${credentialId}`,
+        iconUrl: browser.runtime.getURL("/logo.png"),
+        type: "basic",
+      },
+    });
+
+    return true;
+  };
+
+  deleteAllVerifiableCredentials = async (): Promise<boolean> => {
+    const credentials = await this.getVerifiableCredentialsFromStore();
+    if (credentials.size === 0) {
+      return false;
+    }
+
+    await this.verifiableCredentialsStore.clear();
+    await this.historyService.trackOperation(OperationType.DELETE_ALL_VERIFIABLE_CREDENTIALS, {});
+    await this.notificationService.create({
+      options: {
+        title: "All Verifiable Credentials deleted.",
+        message: `Deleted ${credentials.size} Verifiable Credentials.`,
+        iconUrl: browser.runtime.getURL("/logo.png"),
+        type: "basic",
+      },
+    });
+
+    return true;
+  };
+
+  private insertVerifiableCredentialIntoStore = async (
+    verifiableCredential: VerifiableCredential,
+  ): Promise<boolean> => {
+    /**
+     * We do not accept Verifiable Credentials that do not have a non-empty id.
```

Comment:
If we don't use empty fields maybe we can update the types and make them stricter as we discussed before?

Reply:
Upon doing more research, it seems like some credential schemes (specifically [anonymous credentials](https://hyperledger.github.io/anoncreds-spec/#introduction)) avoid using the id parameter in a credential for linkability purposes. So I think it makes sense for us to use the minimal specification for a Verifiable Credential when storing. That being said, I will have to modify how we index into a stored credential, perhaps by using a hash.

I think enforcing strict types may create some compatibility issues that will bug us in the future.